### PR TITLE
Fix owner of splices in class statements

### LINKED ISSUE #18358: Splices in statements of class definition have the wrong owner.
## Compiler version

If you're not sure what version you're using, run `print scalaVersion` from sbt
(if you're running scalac manually, use `scalac -version` instead).

## Minimized code

```Scala
import scala.quoted.*
inline def generateCode: Unit = ${ testLocalDummyOwner }
def testLocalDummyOwner(using Quotes): Expr[Unit] = '{ trait E { $valInBlock } }
def valInBlock(using Quotes): Expr[Unit] = '{ val x: Int = 2; () }
```

```scala
def test = generateCode
```

## Output

With `-Xcheck-macros`
```scala
1 |def test = generateCode
  |           ^^^^^^^^^^^^tstrapped / Test / testOnly 3s
  |Malformed tree was found while expanding macro with -Xcheck-macros.
  |               |The tree does not conform to the compiler's tree invariants.
  |               |
  |               |Macro was:
  |               |scala.quoted.runtime.Expr.splice[scala.Unit](((contextual$1: scala.quoted.Quotes) ?=> Macro_1$package.testLocalDummyOwner(contextual$1)))
  |               |
  |               |The macro returned:
  |               |{
  |  trait E() extends java.lang.Object {
  |    {
  |      val x: scala.Int = 2
  |      ()
  |    }
  |  }
  |  ()
  |}
  |               |
  |               |Error:
  |               |assertion failed: bad owner; value x has owner trait E, expected was value <local E>
  |owner chain = value x, trait E, method test, package object Test_2$package, package <empty>, package <root>, ctxOwners = value <local E>, value <local E>, value <local E>, value <local E>, trait E, trait E, method test, method test, method test, method test, method test, method test, method test, method test, package object Test_2$package, package <empty>, package <root>, package <root>, package <root>, package <root>, package <root>, package <root>, <none>, <none>, <none>, <none>, <none>
  |               |
  |stacktrace available when compiling with `-Ydebug`
  |               |
  |---------------------------------------------------------------------------------------------------------------------
  |Inline stack trace
  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  |This location contains code that was inlined from Macro_1.scala:2
2 |inline def generateCode: Unit = ${ testLocalDummyOwner }
  |                                ^^^^^^^^^^^^^^^^^^^^^^^^
   ---------------------------------------------------------------------------------------------------------------------
```

It will also fail with `-Ycheck:all`.

Without checks 
```
Unexpected tree in genLoad: DefDef(x,List(List()),TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Int)],Literal(Constant(2)))/class dotty.tools.dotc.ast.Trees$DefDef at: [206..210..220]
java.lang.RuntimeException: Unexpected tree in genLoad: DefDef(x,List(List()),TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Int)],Literal(Constant(2)))/class dotty.tools.dotc.ast.Trees$DefDef at: [206..210..220]
        at dotty.tools.backend.jvm.BCodeHelpers.abort(BCodeHelpers.scala:844)
        at dotty.tools.backend.jvm.BCodeHelpers.abort$(BCodeHelpers.scala:45)
        at dotty.tools.backend.jvm.CodeGen$Impl$.abort(CodeGen.scala:176)
        at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadTo(BCodeBodyBuilder.scala:482)
        at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:303)
        at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genStat(BCodeBodyBuilder.scala:117)
        at dotty.tools.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlockTo$$anonfun$1(BCodeBodyBuilder.scala:1087)
        ...
```

## Expectation
The owner of `x` should be the local dummy `<local E>`.

### PR DESCRIPTION
**Pull request by nicolasstucki**
Fixes #18358

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala [hunk 1]
```diff
@@ -125,6 +125,18 @@ object PickledQuotes {
               val quotedType = evalHole.nn.apply(idx, reifyTypeHoleArgs(args))
               PickledQuotes.quotedTypeToTree(quotedType)
           }
+        case tree @ Template(constr, parents, self, _) if tree.derived.isEmpty =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala:128]: Can tree.derived not be empty here?

File: compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala [hunk 2]
```diff
@@ -125,6 +125,18 @@ object PickledQuotes {
...
+          // Set local dummy as owner when we transform class statements
+          val constr1 = transformSub(constr)
+          val parents1 = transform(tree.parents)
+          val self1 = transformSub(self)
+          val stats1 = flatten(tree.body.mapConserve {
+            case stat: DefTree => transform(stat)
+            case stat =>
+              val localDummy = tree.tpe.termSymbol
+              transform(stat)(using ctx.withOwner(localDummy))
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala:137]: Could TreeMapWithPreciseStatContexts be reused to avoid a special case?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala]: Yes

File: compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala]: I copied it from the `TreeMap` case. Not sure why it is there. It was introduced in fb09e828e2a9cd1320d036e3916555b08fbde52b.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/quoted/PickledQuotes.scala]: Following up in  #18368

**Review by smarter — APPROVED** [review]

**Pull request merged**